src/main/java/be/uantwerpen/minelabs/potion/reactions/Reaction.java

```diff
@@ -1,29 +1,101 @@
 package be.uantwerpen.minelabs.potion.reactions;
 
-import be.uantwerpen.minelabs.potion.GasEntity;
-import be.uantwerpen.minelabs.potion.Molecule;
-import net.minecraft.sound.SoundCategory;
-import net.minecraft.sound.SoundEvents;
+import net.minecraft.block.Block;
+import net.minecraft.block.BlockState;
+import net.minecraft.block.Blocks;
+import net.minecraft.entity.Entity;
+import net.minecraft.entity.LivingEntity;
 import net.minecraft.util.hit.BlockHitResult;
 import net.minecraft.util.hit.EntityHitResult;
 import net.minecraft.util.hit.HitResult;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Box;
 import net.minecraft.util.math.Vec3d;
 import net.minecraft.world.World;
 
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
 import java.util.Objects;
+import java.util.function.Consumer;
 
 public abstract class Reaction {
 
+    private final List<Block> whiteList;
+    private final List<Block> blackList;
+
+    protected Reaction() {
+        this.whiteList = List.of();
+        this.blackList = List.of();
+    }
+
+    protected Reaction(List<Block> whiteList, List<Block> blackList) {
```

Comment:
Can this be changed to work with tags instead? Make a tag for each reaction so blocks can be added in the json instead of the constructor. Depending on the reaction you probably only need a whitelist or only a blacklist so then you can make the most logical choice per reaction instead of having to support both.

Does the list of blocks depend on the substance that caused the reaction? Then the tag will need to be passed through the constructor as well. In any case it will be easier to reuse sets of blocks if multiple substances react the same way and reduce code duplication. I don't think a list of tags is necessary though it might be useful to use existing tags in combination with a custom tag to fill in the gaps.

Reply:
![image](https://user-images.githubusercontent.com/7142721/230767761-34f4423d-7379-46ef-99f9-ae785c4e9cf3.png)
For more info check Tags class in the util package.